File: ruby/resistor-color-trio/resistor_color_trio.rb
=begin
Write your code for the 'Resistor Color Trio' exercise in this file. Make the tests in
`resistor_color_trio_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/resistor-color-trio` directory.
=end

class ResistorColorTrio
  COLORS = %w[black brown red orange yellow green blue violet grey white].freeze

  def initialize(colors)
    raise ArgumentError if colors.size != 3
    raise ArgumentError unless colors.all? { |e| COLORS.include?(e) }

    @colors = colors
  end

  def label
    v = @colors.map { |e| COLORS.index(e) }

    resistor = (v[0] * 10 + v[1]) * 10**v[2]
    resistor = resistor >= 1000 ? "#{resistor / 1000} kiloohms" : "#{resistor} ohms"

    "Resistor value: #{resistor}"
  end
end

File: typescript/resistor-color-trio/resistor-color-trio.ts

enum ResistorValues {
  black = 0,
  brown = 1,
  red = 2,
  orange = 3,
  yellow = 4,
  green = 5,
  blue = 6,
  violet = 7,
  grey = 8,
  white = 9,
}

type Color = keyof typeof ResistorValues;

const Units = [
  {
    base: 10 ** 9,
    value: "gigaohms",
  },
  {
    base: 10 ** 6,
    value: "megaohms",
  },
  {
    base: 10 ** 3,
    value: "kiloohms",
  },
]

export function decodedResistorValue(colors: Color[]): string {
  if (colors.length < 3 || colors.some(r => !Object.prototype.hasOwnProperty.bind(ResistorValues, r))) {
    throw new Error("error color value")
  }

  let [first, second, zero] = colors
  let value = (ResistorValues[first] * 10 + ResistorValues[second]) * 10 ** ResistorValues[zero]

  let unit = Units.find(r => value >= r.base)
  if (unit) {
    return `${value / unit.base} ${unit.value}`
  }
  return `${value} ohms`
}
